python_modules/dagster/dagster/_core/loaders/__init__.py

```diff
@@ -0,0 +1,57 @@
+from abc import abstractmethod
+from typing import Dict, Generic, Iterable, List, Mapping, Optional, Sequence, Type, TypeVar
+
+from dagster._utils.aiodataloader import DataLoader
+
+from ..instance import DagsterInstance
+from ..storage.dagster_run import RunRecord, RunsFilter
+
+T = TypeVar("T")
+
+
+class DagsterDataLoader(
+    DataLoader[str, T],
+    Generic[T],
+):
+    def __init__(self, instance: DagsterInstance):
+        self._instance = instance
+        super().__init__()
+
+
+class LoaderRegistry:
+    """Maintains a mapping of types to their corresponding DataLoaders classes.
+    This mapping is maintained at the process scope, and then used in the request scope
+    to create the appropriate DataLoader for a given type to facilitate request scoped caching.
+    """
+
+    def __init__(self, initial_loaders: Mapping[Type, Type[DagsterDataLoader]]):
+        self._loader_class_map = {**initial_loaders}
+
+    def register_loaders(self, loaders: Mapping[Type, Type[DagsterDataLoader]]) -> None:
+        self._loader_class_map.update(loaders)
+
+    def create_loader_for(self, ttype: Type[T], instance: DagsterInstance) -> DagsterDataLoader[T]:
+        if ttype not in self._loader_class_map:
+            raise KeyError(f"No DataLoader registered for type {ttype}")
+
+        return self._loader_class_map[ttype](instance)
+
+
+class RunRecordLoader(DagsterDataLoader[RunRecord]):
+    async def batch_load_fn(self, keys: Iterable[str]) -> Iterable[Optional[RunRecord]]:
+        result_map: Dict[str, Optional[RunRecord]] = {run_id: None for run_id in keys}
+
+        # this should be replaced with an async DB call
+        records = self._instance.get_run_records(RunsFilter(run_ids=list(result_map.keys())))
+
+        for record in records:
+            result_map[record.dagster_run.run_id] = record
+
+        return result_map.values()
+
+
+DAGSTER_LOADER_REGISTRY = LoaderRegistry(
+    {
+        RunRecord: RunRecordLoader,
+    }
+)
```

Comment:
the line of thinking here is for being able to extend this in internal, might not need to be this mutable / dynamic . Will iterate on this

Reply:
+1 to a registry to enable nicer factory methods